content/BPIP-3.md

```diff
@@ -146,28 +149,113 @@ The following method is added.
+
+Given that now buyer does not directly interact with the protocol, this opens the question, how to ensure protocol exchange guarantees (i.e. possibility to get funds back if something goes wrong). The proposed solution is than now seller needs to provide enough funds to cover both item price and seller deposit. This ensures buyer protection straight away, but it increases a seller's capital requirement a bit. However, this drawback is not to big, since as soon as marketplace exchange happens, seller receives the funds, which can be directly used to cover item price in next sales. Moreover, as soon as some exchange are finalized and funds are released back into seller's pool, those funds can be automatically used to cover future sales.
+
+This proposal might rise some efficiency concerns about the solution efficiency, since minting ERC721 is relatively costly operation. ERC1155 was considered as alternative, but it turned out that with it it's impossible to maintain nice exchange id and token id equivalence. To overcome the problem of ERC721 inefficiency we propose another solution. Since premint always assign token ownership to voucher contract owner, there is no need to explicitly store it's address during the premint. It's enough that only correct transfer events are emitted, while general consistency is achieved by overriding ERC721 `ownerOf`, `transferFrom` and `safeTransferFrom` methods to correctly manage all tokens that were preminted.
 
 ## Backward compatibility
 This specification does not break backward compatibility.
 
+Still it is important that when upgrade is done voucher contracts are upgraded before the protocol is upgraded.
+
 ## Implementation
 ### BosonVoucher
-* Support the tracking of reserved ranges so that exchange id and token id can remain the same.
-* Emit events that look as if vouchers have been minted, but don't store the owner address to conserve the bulk of an actual minting's gas. 
-* In the ownerOf method, if the token id is in a reserved range and has no stored owner, report the contract owner (the seller) as the owner of the voucher. 
-* Add a hook transfer hook that will detect if the voucher is being transferred to the first real owner, and if so, call the commitToPremintedOffer protocol method, which does not require payment.
+
+* Range reservation
+  * Must be done before preminting can happen.
+  * Ranges must be strictly increasing, i.e. start of new range must be greater than current highest range.
+  * Store information about associated offer id, range start and range length.
+  * The tracking of reserved ranges enables that exchange id (in protocol) and token id (voucher contract) can remain the same.
+* Preminting
+  * Can be done in batches. This allows to premint the quantities that could otherwise not be possible because of the block gas limit. Additionally it allows seller, to only partially release vouchers to the market.
```

Comment:
```suggestion
  * Can be done in batches. This allows to premint the quantities that could otherwise not be possible because of the block gas limit. Additionally, it allows sellers to only partially release vouchers to the market.
```